### PR TITLE
add replacer code similar to JSON replacer and (configurable) toJSON

### PR DESCRIPTION
**Pull request by sjpt**
I apologize if I haven't created this request correctly; still very unused to git.
I have changed the base file lib/js-yaml/dumper,js, but not built any dist/js-yaml.js or similar files.
Also, there are just two test cases which are for the moment included inline in dumper.js.

### CODE REVIEW

File: lib/js-yaml/dumper.js [hunk 1]
```diff
@@ -682,7 +684,31 @@ function detectType(state, object, explicit) {
...
+  failing test case, because keys get processed through this as well as values
+    obj = {x:3, y:7, zzz: {q:3, toJSON: function(m) {return this.q || 7}}};
+    repl = (k,v)=> v === 'y' ? k + 993 : v
+    yyy = jsyaml.safeDump(obj, {replacer: repl , localReplacer: 'toJSON', skipInvalid: true})
+    jjj = JSON.stringify(obj, repl)
+    console.log('yyy => ', yyy); console.log('jjj => ', jjj)
+  ***/
+  let object = oobject; // keep original for debug
+  if (typeof object === 'object' && object[state.localReplacer]) {
+    object = object[state.localReplacer].apply(object);
```
Comment (graphicore) [inline lib/js-yaml/dumper.js]: This is not called  correctly:

> JSON.stringify() calls toJSON with one parameter:
>
> * if this object is a property value, the property name
> * if it is in an array, the index in the array, as a string
> * an empty string if JSON.stringify() was directly called on this object


 do it like this:

```js
object_ = object[state.localReplacer](key || '');
```

**NOTE:** we have a problem here, `writeNode` is not aware of the key it is writing.

Also, not each invocation of  `writeNode` writes a value, some invocations just write keys, see the `isKey` parameter. If `isKey` is true, we should never even consider calling any replacer.

File: lib/js-yaml/dumper.js [hunk 2]
```diff
@@ -682,7 +684,31 @@ function detectType(state, object, explicit) {
...
+  }
+  if (state.replacer) {
```
Comment (graphicore) [inline lib/js-yaml/dumper.js]: that's not sufficient to match https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#The_replacer_parameter

File: lib/js-yaml/dumper.js [hunk 3]
```diff
@@ -682,7 +684,31 @@ function detectType(state, object, explicit) {
...
+
+  failing test case, because keys get processed through this as well as values
+    obj = {x:3, y:7, zzz: {q:3, toJSON: function(m) {return this.q || 7}}};
+    repl = (k,v)=> v === 'y' ? k + 993 : v
+    yyy = jsyaml.safeDump(obj, {replacer: repl , localReplacer: 'toJSON', skipInvalid: true})
+    jjj = JSON.stringify(obj, repl)
+    console.log('yyy => ', yyy); console.log('jjj => ', jjj)
+  ***/
+  let object = oobject; // keep original for debug
+  if (typeof object === 'object' && object[state.localReplacer]) {
```
[the remark this answers is not in the document]
Comment (graphicore) [reply lib/js-yaml/dumper.js]: This:

```js
&& object[state.localReplacer]
```

must be rather:

```js
&& Object.prototype.hasOwnProperty.call(object, state.localReplacer)
```

**Comment by rlidwka** [comment]
implemented slightly differently in https://github.com/nodeca/js-yaml/commit/359b264c0fbadfae86caff91879bf40e902df690

**Pull request closed without merging**